### ISSUE TITLE
Incomplete SFTP file copy

### ISSUE THREAD

**Issue by wmiltenburg** [ronf/asyncssh#225]
Hi,

When using assyncSSH for a file copy using SFTP, and without specifying the block_size (default: 16K), my program would only get around 4K of data, while the actual file is around 9K.

After some debugging, I found that the read operation in `_SFTPFileCopier` actually reads only 4K data, while you would expect it would read `block_size` (default 16K). This is caused due to the fact the remote device only returns 4K of data.

For this edge case, wouldn't it be an improvement to add an extra check in SFTPFileCopier to verify that we've actually received the full amount of data? We know how much data the task needs to fetch, so we could validate if we've received the full amount or if we need to try a second time to read the remaining data.

The most simple check is to perform an `if len(data) < size` and perform another read (or multiple) to retrieve the rest of the data.

Adding _SFTPFileCopier for convenience.
```
class _SFTPFileCopier(_SFTPParallelIO):
@asyncio.coroutine
    def run_task(self, offset, size):
        """Copy the next block of the file"""
 
        data = yield from self._src.read(size, offset)
        yield from self._dst.write(data, offset)
 
        if self._progress_handler:
            self._bytes_copied += size
            self._progress_handler(self._srcpath, self._dstpath,
                                   self._bytes_copied, self._total_bytes)
```

The most simple workaround for this is lowering the block_size (in our case it was 4K). Perhaps it's up to the user to specify the best block size, but then it might be better to add a warning logmessage to warn the user that not everything was copied.

What are your thoughts about this case? Would it be an improvement to add such a check?

Tested versions: 1.15.0 - 1.17.0.

With kind regards,
Wouter

**Comment by ronf** [comment]
I'm torn on this. According to the SFTP version 3 spec at https://tools.ietf.org/html/draft-ietf-secsh-filexfer-00, the behavior of FXP_READ is supposed to be as follows:

```
In response to this request, the server will read as many bytes as it
can from the file (up to `len'), and return them in a SSH_FXP_DATA
message.  If an error occurs or EOF is encountered before reading any
data, the server will respond with SSH_FXP_STATUS.  For normal disk
files, it is guaranteed that this will read the specified number of
bytes, or up to end of file.  For e.g. device files this may return
fewer bytes than requested.
```

The SFTPFileCopier code assumes regular files and counts on the fact that the only time a server would return less bytes than requested is when the end of file is encountered. In fact it queries the file size in advance so it doesn't even depend on that, reading only byte ranges within the file size returned. It expects that any errors will return an FXP_STATUS and not an FXP_DATA with fewer bytes.

That said, there is some documentation in the open() method of SFTPClient discussing some known limitations on the maximum size allowed by OpenSSH's SFTP server:

```
           The block_size argument specifies the size of parallel read and
           write requests issued on the file. If set to `None`, each read
           or write call will become a single request to the SFTP server.
           Otherwise, read or write calls larger than this size will be
           turned into parallel requests to the server of the requested
           size, defaulting to 16 KB.

               .. note:: The OpenSSH SFTP server will close the connection
                         if it receives a message larger than 256 KB, and
                         limits read requests to returning no more than
                         64 KB. So, when connecting to an OpenSSH SFTP
                         server, it is recommended that the block_size be
                         set below these sizes.
```

If you have a server which can't return more than 4 KB in a single response, your best bet is to set the block size down to match that.

As you said, it would be possible to work around this by having AsyncSSH issue another read whenever it gets back a partial response instead of trying to fill in zeroes for the missing data. The run_task() function has both the offset and size of the range being read, so figuring out the new read range isn't too hard, but it would be less efficient than setting a smaller block size so that smaller reads are issued in the first place. That way, these reads would happen in parallel and all of the data could be gotten on the first attempt.

Also, there's a question here of what to do with writes. I don't believe there's any way to know how many bytes were written successfully. So, if a system had a hidden upper bound on the amount of data allowed in a write and only wrote a portion of the data requested, there'd be no way to know that and automatically re-issue the write for the rest of that range.

I see that a later version of this spec (starting with https://tools.ietf.org/html/draft-ietf-secsh-filexfer-08) did reword the text on reading to say:

```
      The server MUST not respond with more data than is specified by
      the 'length' parameter.  However, the server MAY respond with less
      data if EOF is reached, an error is encountered, or the servers
      internal buffers can not handle such a large request.
```

However, that draft describes version 6 of the SFTP protocol, which very few clients and servers implement. AsyncSSH is designed to match OpenSSH and most other implementations which are based on version 3 of the protocol.

At a minimum, adding a warning to the current code on incomplete reads is probably a good idea. That way, the user is at least aware that something went wrong, and they could consider trying again with a smaller block size. This warning might trigger in other cases, such as when a file is truncated by some other process in the middle of a copy, but that's also a case where the file contents would potentially be incomplete, ending up with a mix of old & new data.

**Comment by wmiltenburg** [comment]
I can imagine that this is a bit of an edge case. I would like to share some more details about the server implementation, but it's also a bit of a blackbox device I'm dealing with (so I don't have all of the inner details).

I'm still a bit in doubt which of the two solutions I prefer. I think it would be beneficial for the user to at least have an indication if they've received the complete file or a corrupt one (due to setting an improper block_size). However, should it only be a warning message or should an Exception be raised, as the user will probably get a corrupt file (at least, I did)? So I can imagine an Exception would even be better, so that a program (which makes use of asyncssh) can deal with such a situation (lowering the block size, and trying again).

However, as you mentioned, draft 8 outlines that this is allowed behaviour. So, with this in mind, I might actually prefer to perform another read and receive the complete file, as it's expected behaviour. Whereas with an Exception/warning, I would need to start the copy from scratch (with a different block_size), but even then we could end up in the situation again. However, we don't know if we only received a part of the file (as the file_size is only checked at the very beginning of initiating a copy) or  if the original file could have been altered mid-copy. So performing another read, can in some case, end up with different or no data at all.

I'm now also wondering at this point if the device will always return 4K of data (assuming that there's still some left to read from) or that it in some scenarios might even be lower. I'll do some more testing.

At this point I think raising an Exception would be best, because we've copied a file which different properties (file_size) than the original file we expected to copy.

**Comment by ronf** [comment]
I took a closer look at this today, and it turned out to be simpler than I expected to allow the code to automatically adapt if it gets back a partial response to a read, even when performing parallel I/O on multiple sections of the file at once. So, I've gone ahead and added this support in commit 2410048 in the "develop" branch.

As part of this change, I also added code to raise an "Unexpected EOF during file copy" SFTPError if one of the reads gets back no data at all. This should only ever happen in cases where a source file is changed (or more specifically truncated) during the copy. This check won't catch all cases where a source file is changed, but if a read does happen to return EOF, it'll be treated like other file I/O errors (including allowing an error_handler to be set to skip to the next file if multiple files are being copied).

One nice thing about the approach here is that it doesn't need to know the exact block size the server supports. Each time a read returns less than the expected amount, it automatically reissues the read for the rest, which should maximize the amount of data it gets back on each read to whatever the server is willing to provide. These retried reads are sequential, but multiple such retries can happen in parallel based on the block size requested in the read or copy request.

If you have a chance to try this change, please let me know what you think!

**Comment by wmiltenburg** [comment]
This looks really good. Thanks!

I've reviewed the code and this should solve the use-case I'm experiencing. I'm still locked out of the VPN, so I will try it tomorrow. Sorry for the delay.

**Comment by ronf** [comment]
Thanks - this is now in the "develop" branch as commit be27d04.

**Comment by wmiltenburg** [comment]
Just tested it and it works. Thanks!

**Comment by wmiltenburg** [comment]
>These retried reads are sequential, but multiple such retries can happen in parallel based on the block size requested in the read or copy request.

Would it also be a good idea to write a log message if you've received a partial read, so that the user could update the block_size for subsequent file copies? If I understand correctly, the reads in a task a sequential, but if the user set the block_size too big, you would end up with multiple sequential reads that could otherwise be parallelised. By looking at the logs (info or debug) the user has an indication that it should optimize the block size.

**Comment by ronf** [comment]
While the messages don't specifically call out partial reads, there are already level 1 debug messages which can be used to see if this is happening. They clearly show the read request size and response size. Here's an example which shows how OpenSSH has an undocumented limit of 64 KB on reads. I read a 1 MB file from an OpenSSH server with a block size of 256 KB, and got the following debug output:

```
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 262144 bytes at offset 0 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 262144 bytes at offset 262144 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 262144 bytes at offset 524288 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 262144 bytes at offset 786432 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 0 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 262144 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 524288 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 786432 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 196608 bytes at offset 65536 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 196608 bytes at offset 327680 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 196608 bytes at offset 589824 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 196608 bytes at offset 851968 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 65536 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 327680 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 589824 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 851968 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 131072 bytes at offset 131072 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 131072 bytes at offset 393216 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 131072 bytes at offset 655360 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 131072 bytes at offset 917504 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 131072 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 393216 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 655360 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 917504 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 65536 bytes at offset 196608 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 65536 bytes at offset 458752 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 65536 bytes at offset 720896 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending read for 65536 bytes at offset 983040 in handle 00000000
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 196608 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 458752 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 720896 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received 65536 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending write for 65536 bytes at offset 983040 in handle 00000001
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received OK
```

Note how it issues 4 parallel 256 KB reads, and when those only return 64 KB, it issues 4 parallel 192 KB reads. This repeats with 128 KB reads and finally 64 KB reads that complete the transfer. There are 16 reads total, which would be the same number as you'd get if you explicitly requested a block size of 64 KB to start with.

While it's true that requesting a smaller block size in this case would have resulted in more parallelism since the file was so small, that's true independent of the block size the server was capable of supporting. If you request a block size bigger than the file size divided max requests (which defaults to 128), you'll be reducing the parallelism.

On the flip size, where this really matters is on larger files, and there you'll end up with the same number of parallel reads and the same total number of reads regardless of whether you choose a block size the server can support or not. The only thing that changes is the ordering of the reads.

Consider a simple example where I'll use powers of 10 to make the math clearer. Take a file size of 10,000 bytes with a requested block size of 100 and max requests of 10, with a server that only supports a block size of 10. When you request block size 100, you'll see reads issued with offsets like 0, 100, 200, ..., 900 in the first pass. As these reads complete, you'll see reads of 10, 110, 210, ..., 910. This will continue until you get to 90, 190, 290, ..., 990. At that point, you'll start to see it move on to 1000, 1100, 1200, ..., 1900 and the pattern will repeat. In the end, you'll end up with a total of 1000 10-byte reads, and you'll have had 10 parallel reads outstanding for the duration of the transfer.

With a requested block size of 10, you'll still see the same 1000 10-byte reads and 10 parallel requests. The only difference will be that the ordering will look something like 0, 10, 20, ..., 90 in the first set, and then 100, 110, 120, ..., 190 as that group completes, continuing to increment by 10 each time a read completes.

In the end, I'd expect these two cases to perform pretty much the same, and I'm hesitant to have an INFO-level message reporting partial reads as that's a lot of extra log output compared to what shows up today (which includes nothing about any individual reads or writes). If someone really wants this level of detail, the existing level 1 debug messages should provide it.

That said, thanks for having me check on this -- I actually found a bug that I would have missed if I didn't pay close attention to the read block sizes in the debug output! I'll have that checked in shortly.

**Comment by ronf** [comment]
This improvement is now available in AsyncSSH 1.17.1.

**Comment by wmiltenburg** [comment]
Thanks for your clear explanation!
